lib/engine/game/g_18_india/step/corporate_buy_shares_company.rb

```diff
@@ -0,0 +1,183 @@
+            return false if @game.in_ipo?(company) && !can_buy_comp_from_ipo?(entity, company)
+
+            entity.cash >= company.value
+          end
+
+          def process_corporate_buy_company(action)
+            entity = action.entity
+            company = action.company
+            price = action.price
+            owner = company.owner
+            what = company.type == :share ? "a share of #{company.name}" : "Private #{company.name}"
+
+            raise GameError, "Cannot buy #{company.name} from #{owner.name}" if owner&.corporation? || owner&.player?
+
+            if @game.in_ipo?(company)
+              row, _index = @game.ipo_row_and_index(company)
+              location = 'IPO Row: ' + (row + 1).to_s
+              @game.ipo_remove(row, company)
+            elsif owner == @game.bank
+              location = 'the Bank'
+              @game.bank.companies.delete(company)
+            end
+            @log << "#{current_entity.name} buys #{what} from #{location} for #{@game.format_currency(price)}"
+
+            case company.type
+            when :share
+              share = company.treasury
+              corp = share.corporation
+              already_floated = corp.floated?
+              # transfer share and send payment to corporation
+              bundle = ShareBundle.new(share)
+              @game.share_pool.transfer_shares(bundle, entity, spender: entity, receiver: corp, price: price)
+              if corp.floatable && corp.floated? && (already_floated == false)
+                @game.float_corporation(corp)
+                maybe_place_home_token(corp)
+              end
+              @round.corporations_bought[entity] << corp
+            when :private, :bond
+              company.owner = entity
+              entity.companies << company
+              entity.spend(price, @game.bank)
+              @round.corporations_bought[entity] << company
+            end
+            debugging_log('Process > Corporate Buy Company')
+          end
+
+          # ----- methods for buying shares -----
+
+          # modified to only check market
+          def can_buy_any?(entity)
+            can_buy_any_from_market?(entity)
+          end
+
+          # modified to allow buying self from market
+          def can_buy?(entity, bundle)
+            return if bought?(entity)
+            return unless bundle
+            return unless bundle.buyable
+            return unless bundle.corporation.ipoed
+            return if bundle.presidents_share
+
+            entity.cash >= bundle.price
+          end
+
+          # modified for debugging info
+          def process_corporate_buy_shares(action)
+            buy_shares(action.entity, action.bundle)
+            @round.corporations_bought[action.entity] << action.bundle.corporation
+            pass! unless can_buy_any?(action.entity)
+            debugging_log('Process > Corporate Buy Shares')
+          end
```

Comment:
It doesn't look like you're changing any logic here, is this ok instead?

```suggestion
          def process_corporate_buy_shares(action)
            super
            debugging_log('Process > Corporate Buy Shares')
          end
```